src/weekly/utils.rs

```diff
@@ -0,0 +1,181 @@
+//! Utilities for weekly-time related logic.
+
+use crate::weekly::{MinuteInWeek, MAX_WEEKLY_MINS};
+use chrono::{DateTime, Utc, Weekday};
+use failure::{bail, ensure, format_err, Fallible};
+use std::time::Duration;
+
+/// Convert datetime to minutes since beginning of week.
+pub(crate) fn datetime_as_weekly_minute(datetime: &DateTime<Utc>) -> MinuteInWeek {
+    use chrono::{Datelike, Timelike};
+
+    let weekday = datetime.weekday();
+    // SAFETY: hour() always <= 23.
+    let hour = datetime.hour() as u8;
+    // SAFETY: minutes() always <= 59.
+    let minute = datetime.minute() as u8;
+
+    time_as_weekly_minute(weekday, hour, minute)
+}
+
+/// Convert a point in weekly-time to minutes since beginning of week.
+pub(crate) fn time_as_weekly_minute(day: chrono::Weekday, hour: u8, minute: u8) -> MinuteInWeek {
+    let hour_minutes = u32::from(hour.min(23)).saturating_mul(60);
+    let day_minutes = day
+        .num_days_from_monday()
+        .saturating_mul(24)
+        .saturating_mul(60);
+    let weekly_minute = day_minutes
+        .saturating_add(hour_minutes)
+        .saturating_add(u32::from(minute.min(59)));
+
+    assert!(weekly_minute < MAX_WEEKLY_MINS);
+    weekly_minute
+}
+
+/// Check duration for a sane lower and upper bound (whole week).
+pub(crate) fn check_duration(length: Duration) -> Fallible<Duration> {
+    // Whole week duration, in seconds.
+    const MAX_WEEKLY_SECS: u64 = MAX_WEEKLY_MINS as u64 * 60;
+
+    if length.as_secs() > MAX_WEEKLY_SECS {
+        failure::bail!("length longer than a week")
+    };
+    if length.as_secs() == 0 {
+        failure::bail!("zero-length duration")
+    };
+
+    Ok(length)
+}
+
+/// Parse a week day string (English names).
+pub(crate) fn weekday_from_string(input: &str) -> Fallible<Weekday> {
+    let day = match input.to_lowercase().as_str() {
+        "mon" | "monday" => Weekday::Mon,
+        "tue" | "tuesady" => Weekday::Tue,
+        "wed" | "wednesday" => Weekday::Wed,
+        "thu" | "thursday" => Weekday::Thu,
+        "fri" | "friday" => Weekday::Fri,
+        "sat" | "saturday" => Weekday::Sat,
+        "sun" | "sunday" => Weekday::Sun,
+        _ => bail!("unrecognized week day: {}", input),
+    };
+
+    Ok(day)
+}
+
+/// Parse a time string (in 24h format).
```

Comment:
Would help to have a sample time string here.

Reply:
Compiler checked examples in docstrings FTW!